exp/mutators/azureasomanagedmachinepool.go

```diff
@@ -0,0 +1,155 @@
+				agentPool = u
+				agentPoolPath = fmt.Sprintf("spec.resources[%d]", i)
+				break
+			}
+		}
+		if agentPool == nil {
+			return reconcile.TerminalError(ErrNoManagedClustersAgentPoolDefined)
+		}
+
+		if err := setAgentPoolOrchestratorVersion(ctx, machinePool, agentPoolPath, agentPool); err != nil {
+			return err
+		}
+
+		if err := reconcileAutoscaling(agentPool, machinePool); err != nil {
+			return err
+		}
+
+		if err := setAgentPoolCount(ctx, machinePool, agentPoolPath, agentPool); err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+func setAgentPoolOrchestratorVersion(ctx context.Context, machinePool *expv1.MachinePool, agentPoolPath string, agentPool *unstructured.Unstructured) error {
+	_, log, done := tele.StartSpanWithLogger(ctx, "mutators.setAgentPoolOrchestratorVersion")
+	defer done()
+
+	if machinePool.Spec.Template.Spec.Version == nil {
+		return nil
+	}
+
+	k8sVersionPath := []string{"spec", "orchestratorVersion"}
+	capiK8sVersion := strings.TrimPrefix(*machinePool.Spec.Template.Spec.Version, "v")
+	userK8sVersion, k8sVersionFound, err := unstructured.NestedString(agentPool.UnstructuredContent(), k8sVersionPath...)
+	if err != nil {
+		return err
+	}
+	setK8sVersion := mutation{
+		location: agentPoolPath + "." + strings.Join(k8sVersionPath, "."),
+		val:      capiK8sVersion,
+		reason:   fmt.Sprintf("because MachinePool %s's spec.template.spec.version is %s", machinePool.Name, *machinePool.Spec.Template.Spec.Version),
+	}
+	if k8sVersionFound && userK8sVersion != capiK8sVersion {
+		return Incompatible{
+			mutation: setK8sVersion,
+			userVal:  userK8sVersion,
+		}
+	}
+	logMutation(log, setK8sVersion)
+	return unstructured.SetNestedField(agentPool.UnstructuredContent(), capiK8sVersion, k8sVersionPath...)
+}
+
+func reconcileAutoscaling(agentPool *unstructured.Unstructured, machinePool *expv1.MachinePool) error {
+	autoscaling, _, err := unstructured.NestedBool(agentPool.UnstructuredContent(), "spec", "enableAutoScaling")
+	if err != nil {
+		return err
+	}
+
+	// Update the MachinePool replica manager annotation. This isn't wrapped in a mutation object because
+	// it's not modifying an ASO resource and users are not expected to set this manually. This behavior
+	// is documented by CAPI as expected of a provider.
+	replicaManager, hasReplicaManager := machinePool.Annotations[clusterv1.ReplicasManagedByAnnotation]
+	if autoscaling {
+		if !hasReplicaManager {
+			if machinePool.Annotations == nil {
+				machinePool.Annotations = make(map[string]string)
+			}
+			machinePool.Annotations[clusterv1.ReplicasManagedByAnnotation] = infrav1exp.ReplicasManagedByAKS
+		} else if replicaManager != infrav1exp.ReplicasManagedByAKS {
+			return fmt.Errorf("failed to enable autoscaling, replicas are already being managed by %s according to MachinePool %s's %s annotation", replicaManager, machinePool.Name, clusterv1.ReplicasManagedByAnnotation)
+		}
+	} else if !autoscaling && replicaManager == infrav1exp.ReplicasManagedByAKS {
```

Comment:
Can we add a comment here to indicate that this implements the required reconciliation logic to enable mutating from "node pool is autoscaler-enabled" to "node pool is not autoscaler-enabled"?

Reply:
Not sure I exactly get what you mean, but I took a stab here.

Reply:
ya this works